### DISCUSSION TITLE
Mqtt Subscription

### DISCUSSION THREAD

**Discussion by snico2** [roleoroleo/yi-hack-Allwinner-v2#278] [Q&A] [answered]
Hi roleo, the Hack, through web services, make available some functions such as "Get recorded videos".

Is it possible to make these functions also callable through the mqtt protocol?

Currently the Hack has the mqtt subscription only msg the "IPC_CMD" commands through the mqtt_set_config.sh script.

I think we should extend this script to manage the other msg subscripts (get_recorded video), then properly recall the scripts already present in cgi-bin such as getlastrecordedvideso.sh.

I tried to invoke the getlastrecordedvideso.sh script but I can't set the QUERY_STRING variable correctly (for example oldness=1&type=2).

Sorry for my bad english but my native langua is italian

snico

**Comment by snico2** [comment, marked as answer]
Hi, fixed problem with QUERY_STRING variable, 

I created the callit.sh script that call any script inside "yi-hack\www\cgi-bin" directory and set QUERY_STRING varaible.

#!/bin/sh
SCRIPT_NAME=$1
QUERY_STRING=$2
YI_HACK_PREFIX="/tmp/sd/yi-hack"
SCRIPT_PATH=$YI_HACK_PREFIX"/www/cgi-bin/"$SCRIPT_NAME".sh"
export QUERY_STRING
"$SCRIPT_PATH"

Example:
./callit.sh  getlastrecordedvideo "type=2&oldness=1" 

Now i must create the script "mqtt_get_attribute.sh" as "mqtt_set_config.sh" like and subscribe mqtt msg for get attibute ...

**Comment by snico2** [comment]
Hi Roleo, i need your help for understand mqtt subscription work with firmware

I have configurate mqtt and it work very well for publish message (the process mqttv4 is active)

When i set "Topics Prefix for remote configuration" it dont't work.

Ex. I set "Topics Prefix = yicam-config" and then I publish (from mosquitto windows) the msg:

	D:\mosquitto>mosquitto_pub -h 192.168.1.10 -p 1883 -u nico -P nico -t yicam-config/system/rtsp -m no
	
but the file /tmp/sd/yi-hack/etc/system.conf  report always "RTSP=yes"

The script "yi-hack/script/mqtt_advertise/mqtt_set_config.sh" work well (on camera setting attribute as LED, IR, ecc) only when mqttv4 process is inactive.

Any solution ?
Thanks

**Reply by roleoroleo** [reply]
> Hi Roleo, i need your help for understand mqtt subscription work with firmware
> 
> I have configurate mqtt and it work very well for publish message (the process mqttv4 is active)
> 
> When i set "Topics Prefix for remote configuration" it dont't work.
> 
> Ex. I set "Topics Prefix = yicam-config" and then I publish (from mosquitto windows) the msg:
> 
> ```
> D:\mosquitto>mosquitto_pub -h 192.168.1.10 -p 1883 -u nico -P nico -t yicam-config/system/rtsp -m no
> ```
> 
> but the file /tmp/sd/yi-hack/etc/system.conf report always "RTSP=yes"

I tested it now and it works with my y21ga.
Check your config and your mqtt server.

**Reply by snico2** [reply]
Sorry, can you post your test topic & paylod ?
After message published your file system.conf is updated?

I tried with my r30gb & h201c
I use Mosquitto server on Windows and also a mosquitto_sub (on windows for check)

I must enable something for subscribe message?
From mosquitto server log i see that yicam don't subscribe none, it publish only

Thanks

**Comment by snico2** [comment]
Hi roleo, i found the problem.

The process mqtt-config don't is active.

mqtt-config is started from script system.sh  ...

ipc_multiplexer &
sleep 1
if [[ $(get_config MQTT) == "yes" ]] ; then
    mqttv4 &
    mqtt-config &
fi

I cannot read log file of mqtt-config at boot.

But if start manually the script system.sh i see ...

Starting mqttv4 v0.0.3
key: MQTT_PREFIX_CONFIG | value: yicam-config
Unrecognized config.
Trying to connect... Ignoring key: MQTT_PREFIX - value: yahas
Ignoring key: TOPIC_BIRTH_WILL - value: virtual/reply/camtest/camtest/status
Ignoring key: TOPIC_AI_HUMAN_DETECTION - value: virtual/reply/camtest/camtest/human
Ignoring key: TOPIC_MOTION - value: virtual/reply/camtest/camtest/motion
Ignoring key: TOPIC_MOTION_IMAGE - value: virtual/reply/camtest/camtest/image
....

Why mqtt-config don't start at boot ?

**Comment by snico2** [comment]
Another little step ...

I updated system.sh script with ...  

if [[ $(get_config MQTT) == "yes" ]] ; then
    mqttv4 &
	sleep 3
    mqtt-config >> "/tmp/sd/yi-hack/mqtt-config.log" 2>&1 &
fi

Now i can see log file ...

root@yitest:~# cat mqtt-config.log
Ignoring key: MQTT_PREFIX - value: yahas
Ignoring key: TOPIC_BIRTH_WILL - value: virtual/reply/camtest/camtest/status
Ignoring key: TOPIC_AI_HUMAN_DETECTION - value: virtual/reply/camtest/camtest/human
Ignoring key: TOPIC_MOTION - value: virtual/reply/camtest/camtest/motion
Ignoring key: TOPIC_MOTION_IMAGE - value: virtual/reply/camtest/camtest/image
Ignoring key: TOPIC_MOTION_FILES - value: motion_files
Ignoring key: TOPIC_BABY_CRYING - value: baby_crying
Ignoring key: TOPIC_SOUND_DETECTION - value: virtual/reply/camtest/camtest/sound
Ignoring key: BIRTH_MSG - value: {"operation":"ENABLED"}
Ignoring key: WILL_MSG - value: {"operation":"DISABLED"}
Ignoring key: AI_HUMAN_DETECTION_START_MSG - value: {"alert":"START"}
Ignoring key: AI_HUMAN_DETECTION_STOP_MSG - value: {"alert":"STOP"}
Ignoring key: MOTION_START_MSG - value: {"alert":"START"}
Ignoring key: MOTION_STOP_MSG - value: {"alert":"STOP"}
Ignoring key: BABY_CRYING_MSG - value: crying
Ignoring key: SOUND_DETECTION_MSG - value: {"alert":"SOUND"}
Ignoring key: MQTT_KEEPALIVE - value: 120
Ignoring key: MQTT_QOS - value: 1
Ignoring key: MQTT_RETAIN_BIRTH_WILL - value: 1
Ignoring key: MQTT_RETAIN_AI_HUMAN_DETECTION - value: 1
Ignoring key: MQTT_RETAIN_MOTION - value: 1
Ignoring key: MQTT_RETAIN_MOTION_IMAGE - value: 1
Ignoring key: MQTT_RETAIN_MOTION_FILES - value: 1
Ignoring key: MQTT_RETAIN_BABY_CRYING - value: 1
Ignoring key: MQTT_RETAIN_SOUND_DETECTION - value: 1
Ignoring key: MQTT_RETAIN_SOUND_DETECTION - value: 1
Unable to connect to the broker

mqtt-config don't connect to mqtt broker (why? boh !!)

After boot i can start manually system.sh ...

root@yitest:~# cat test6.sh
#!/bin/sh

YI_HACK_PREFIX="/tmp/sd/yi-hack"

PATH=$PATH:$YI_HACK_PREFIX/bin:$YI_HACK_PREFIX/usr/bin
LD_LIBRARY_PATH=$YI_HACK_PREFIX/lib:$LD_LIBRARY_PATH

mqtt-config
root@yitest:~# ./test6.sh
Ignoring key: MQTT_PREFIX - value: yahas
Ignoring key: TOPIC_BIRTH_WILL - value: virtual/reply/camtest/camtest/status
Ignoring key: TOPIC_AI_HUMAN_DETECTION - value: virtual/reply/camtest/camtest/human
Ignoring key: TOPIC_MOTION - value: virtual/reply/camtest/camtest/motion
Ignoring key: TOPIC_MOTION_IMAGE - value: virtual/reply/camtest/camtest/image
Ignoring key: TOPIC_MOTION_FILES - value: motion_files
Ignoring key: TOPIC_BABY_CRYING - value: baby_crying
Ignoring key: TOPIC_SOUND_DETECTION - value: virtual/reply/camtest/camtest/sound
Ignoring key: BIRTH_MSG - value: {"operation":"ENABLED"}
Ignoring key: WILL_MSG - value: {"operation":"DISABLED"}
Ignoring key: AI_HUMAN_DETECTION_START_MSG - value: {"alert":"START"}
Ignoring key: AI_HUMAN_DETECTION_STOP_MSG - value: {"alert":"STOP"}
Ignoring key: MOTION_START_MSG - value: {"alert":"START"}
Ignoring key: MOTION_STOP_MSG - value: {"alert":"STOP"}
Ignoring key: BABY_CRYING_MSG - value: crying
Ignoring key: SOUND_DETECTION_MSG - value: {"alert":"SOUND"}
Ignoring key: MQTT_KEEPALIVE - value: 120
Ignoring key: MQTT_QOS - value: 1
Ignoring key: MQTT_RETAIN_BIRTH_WILL - value: 1
Ignoring key: MQTT_RETAIN_AI_HUMAN_DETECTION - value: 1
Ignoring key: MQTT_RETAIN_MOTION - value: 1
Ignoring key: MQTT_RETAIN_MOTION_IMAGE - value: 1
Ignoring key: MQTT_RETAIN_MOTION_FILES - value: 1
Ignoring key: MQTT_RETAIN_BABY_CRYING - value: 1
Ignoring key: MQTT_RETAIN_SOUND_DETECTION - value: 1
Ignoring key: MQTT_RETAIN_SOUND_DETECTION - value: 1
Callback, connected with return code rc=0

Now it is Ok

**Comment by snico2** [comment]
Good news. 
I found workaround solution.
Inside system.sh script the sleep time must be increased to at least 4 seconds 

if [[ $(get_config MQTT) == "yes" ]] ; then
mqttv4 &
sleep 5
mqtt-config >> "/tmp/sd/yi-hack/mqtt-config.log" 2>&1 &
fi

With sleep 3 second mqtt-config start fails 
With sleep 4 second mqtt-config start is ok

For prudence better set sleep to 5 

Test done on YI Dome Guard r30gb model.

**Comment by roleoroleo** [comment]
Wow...
This is very strange.
Never seen a problem like this.
I will check if I can add a sleep.
